streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -714,19 +720,34 @@ public boolean process(final long wallClockTime) {
             if (recordInfo.queue().size() == maxBufferedSize) {
                 mainConsumer.resume(singleton(partition));
             }
-        } catch (final StreamsException e) {
-            throw e;
+
+            record = null;
+        } catch (final TimeoutException timeoutException) {
+            if (!eosEnabled) {
+                throw timeoutException;
```

Comment:
It seems like KIP-572 should retry only when it's safe to do so. Under ALOS, it's not safe to retry, so we should just crash here, right?

Reply:
If I'm reading the code right, this would also mean we don't need to mess with caching the StampedRecord, which would be nice.

Reply:
@vvcephei what do you mean by "crash" here? Throw TaskCorrutpedException?

Reply:
Why would it not be safe to retry the commit for at-least-once? -- The indention of the PR was to retry for this case.

Reply:
Thanks, all, sorry for the delay. I meant "crash" as in, crash the thread. This is what happens today, right? We throw the TimeoutException, convert it to a StreamsException, and throw it up to the top level, killing the thread?

What I am thinking is that it's silly to just crash a thread when it has gotten a timeout in `poll()` or in checking `partitions()` because we can harmlessly retry those operations. Along those lines, retrying `commit` should be safe as well.

But it looks like this PR is not retrying just `commit`, but is retrying the entire processing of a record. Is that a wrong conclusion?

Retrying "process a record" is definitely not "safe" or "harmless" it will absolutely result in producing incorrect results, therefore we should not do it.

Reply:
> But it looks like this PR is not retrying just commit, but is retrying the entire processing of a record. Is that a wrong conclusion?
>
> Retrying "process a record" is definitely not "safe" or "harmless" it will absolutely result in producing incorrect results, therefore we should not do it.

Well, overall I agree, but if we crash the thread with at-least-once, a we rebalance, the other thread will reprocess based on the last committed offset, so the net result is, that a rebalance might reprocess even _more_ records -- if we handle the timeout, we only try to reprocess a single record.